### PR TITLE
Edu/zero upper bits1

### PR DESCRIPTION
**Pull request by ed255**
- Emulator:
  - Fix bounds check for `imm` instruction inputs: `2^x` is `1 << x` and not `2 << x`
  - Rearange `imm` bounds checks to keep `imm` in the middle
  - Fix bounds check in `lui` considering that we already have the value shifted by 12
- Simulator:
  - Remove `wp1_to_w` and `wx2_to_w` table simulation in favor of structured table simulation
  - Introduce subtables required for range checks, zeroing upper bits removal, and zeroing lower bit
  - Introduce `zero_upper_bits`, `zero_upper_bits_lower_bit` and `range_nbits` structured tables using the subtables from the previous point
  - Annotate errata from Jolt paper when our implementation differs due to Jolt paper errors
  - Fix `eq` structured table bug in for loop and bit slicing
  - Simulate `lui` and `addi`
  - Use new "range-like" structured tables in `add`, `sub`, `jal` and `jalr`
  - Handle negative `imm` in `beq` (I think the nested if will need to be translated into a polynomial expression to be handled via constraint)
- Test:
  - Add tests for simulator & emulator of the instructions `addi`, `lui`, `beq`, `jal` and `jalr`
- Utils (new module):
  - Add helper type wrappers to display `ark_ff` field elements nicely with `Display` (useful for debug printing vectors of field elements)

### CODE REVIEW

File: src/simulator.rs [hunk 1]
```diff
@@ -302,52 +431,75 @@ impl<F: PrimeField> Simulator<F> {
...
-        }
+        let condition = LookupTables::eq(W, C, z);
+        // Simulate the encoding of the instruction from Jolt (where opflag=positive)
+        let (positive, imm) = {
+            if imm >= 0 {
+                (F::one(), F::from(imm as u64))
+            } else {
+                (F::zero(), F::from((-imm) as u64))
+            }
+        };
```
Comment (jeong0982) [inline src/simulator.rs:463]: Or we can get this by function parameter. In the Jolt paper, they make instruction should have the sign of `imm` value.

File: src/simulator.rs [hunk 2]
```diff
@@ -97,6 +89,53 @@ impl<F: Arithmetic> SubTableMLE<F> {
...
+    }
+
+    // Return the input if it uses up to cuttoff bits, or 0 otherwise.
+    fn range_remainder(cutoff: usize, x: &[F]) -> F {
+        let mut result = F::zero();
+        for i in 0..x.len() {
+            if i < cutoff {
+                result += x[i].clone() * f_pow::<F>(2, i);
+            } else {
+                result *= F::one() - x[i].clone();
```
Comment (jeong0982) [inline src/simulator.rs:108]: Why is it doing `1 - x[i]` when `i >= cutoff`?

File: src/simulator.rs (hunk 1 above)
Comment (ed255) [reply src/simulator.rs:463]: You're right!  I've opened an issue for that https://github.com/ed255/riscu-jolt/issues/21
Then all the simulated instructions should work with the `JoltInstruction` type which includes the `opflags`.

File: src/simulator.rs (hunk 2 above)
Comment (jeong0982) [reply src/simulator.rs:108]: Now got it. Missed `*`

**Review by jeong0982 — APPROVED** [review]

**Pull request merged**